operator/pkg/central/extensions/reconcile_central_db_password.go

```diff
@@ -26,105 +27,165 @@ const (
 }
 
-func (r *reconcileCentralDBPasswordExtensionRun) validateSecretData(data types.SecretDataMap, _ bool) error {
-	password, err := passwordFromSecretData(data)
+func obtainPasswordFromReferencedSecret(ctx context.Context, client ctrlClient.Client, referencedSecretKey ctrlClient.ObjectKey) (password string, err error) {
+	// get the referenced secret
+	referencedSecret := new(coreV1.Secret)
+	if err := client.Get(ctx, referencedSecretKey, referencedSecret); err != nil {
+		return "", errors.Wrapf(err, "failed to get spec.central.db.passwordSecret %q", referencedSecretKey.Name)
+	}
+
+	// get the password from the referenced secret
+	password, err = getAndValidatePasswordFromReferencedSecret(referencedSecret)
 	if err != nil {
-		return errors.Wrap(err, "validating existing secret data")
+		return "", errors.Wrapf(err, "reading central db password from secret %s", canonicalCentralDBPasswordSecretName)
 	}
-	if r.password != "" && r.password != password {
-		return errors.New("existing password does not match expected one")
+
+	return password, nil
+}
+
+func updateCentralDBPasswordSecretDataIfNeeded(secret *coreV1.Secret, password string, hasReferencedSecret bool) bool {
+	shouldUpdatePassword := false
+	passwordIsEmpty := secret.Data == nil || len(secret.Data[centralDBPasswordKey]) == 0
+	passwordsAreDifferent := secret.Data == nil || string(secret.Data[centralDBPasswordKey]) != password
+
+	if passwordIsEmpty || hasReferencedSecret && passwordsAreDifferent {
+		shouldUpdatePassword = true
 	}
-	// The following assignment shouldn't have any consequences, as a successful validation should prevent generation
-	// from being invoked, but better safe than sorry (about clobbering a user-set password).
-	r.password = password
-	return nil
+
+	if shouldUpdatePassword {
+		secret.Data = map[string][]byte{
+			centralDBPasswordKey: []byte(password),
+		}
+	}
+
+	return shouldUpdatePassword
 }
 
-func (r *reconcileCentralDBPasswordExtensionRun) generateDBPassword() (types.SecretDataMap, error) {
-	if r.password == "" {
-		r.password = renderer.CreatePassword()
+func unsetCentralDBPasswordSecretOwnerReferenceIfNeeded(c *platform.Central, secret *coreV1.Secret) bool {
+	// make sure that the secret owner reference is unset. This is to ensure that PVCs which are not deleted
+	// when Centrals are deleted do not have their passwords deleted.
+	shouldUpdateOwnerReference := false
+	centralOwnerRef := v1.NewControllerRef(c, c.GroupVersionKind())
+	centralGK := c.GroupVersionKind().GroupKind()
+	for i := len(secret.OwnerReferences) - 1; i >= 0; i-- {
+		ownerRef := secret.OwnerReferences[i]
+		ownerGV, err := schema.ParseGroupVersion(ownerRef.APIVersion)
+		if err != nil {
+			continue
+		}
+		ownerGK := ownerGV.WithKind(ownerRef.Kind).GroupKind()
+		if ownerRef.UID == centralOwnerRef.UID &&
+			ownerRef.Name == centralOwnerRef.Name &&
+			ownerGK == centralGK {
+			secret.OwnerReferences = append(secret.OwnerReferences[:i], secret.OwnerReferences[i+1:]...)
```

Comment:
In the first loop iteration, `[i+1]` will be an out-of-bounds read, no? What am I missing?

Reply:
Yeah that boggles me. I was pretty sure this was the canonical way of "splicing" a slice, but I made some tests and it seems to not be the case. Used filtering instead